### ISSUE TITLE
Contributors wanted?

### ISSUE THREAD

**Issue by ibizaman** [stefanboere/ferp-hs#1]
Hi! What you're trying to accomplish with this project is something that resonates with me. I'm also trying to build a scaffolding using reflex and servant in order to create other apps faster. But you're waaaay further than where I'm at. I especially like you took the security aspect into account - I see keycloak in there :+1: 

Would you be willing to have other contributors help you? Me, in this case?

**Comment by stefanboere** [comment]
Yes, of course, that would be awesome! I made you contributor of this project, so you have direct access now. Of course no strings attached, spend your time wisely and only spend time in this repo if it is worthwile for you :). 

If you find things that are unclear, please let me know and then I can maybe add some more documentation in those places, etc.

**Comment by ibizaman** [comment]
Well... I'd like already one piece of advice. How do you build/run the different projects?

**Comment by stefanboere** [comment]
Ah yes, sorry, this was not there yet indeed. I added something in the readme, but still it can be a bit daunting maybe. 
It would be helpful if you understand and use flakes, although maybe it is possible to run it without flakes (I added the flake-compat files). 

Basically you need to:
- Install postgresql and create a new database
- Install keycloak
     - Create a new realm
     - Create a new client, which is used by backend to connect to keycloak
     - Create a user for yourself to be able to login
     - Create the role `administrator` and add your user to this role. 
- Edit `backend/config.dhall`, `backend-api/config.dhall` and `frontend/config.json` according to your installation above.
- Build all the static files (this will take a while). The `backend` needs this folder and serves this folder. 
      - With flakes: `nix build .\#frontend-min -o result-frontend-min`
      - Without flakes: `nix-build -A packages.x86_64-linux.frontend-min -o result-frontend-min`
- Run `backend-api` and `backend` 
      - With cabal (might not work currently due to some GLIBC version mismatch error): 
             - Enter the development shell (`nix-shell` or `nix develop`)
             - `cd backend-api && cabal run backend-api:exes` and  `cd backend && cabal run backend:exes`
      - With nix flakes
             - `cd backend-api && nix build .\#backend-api` and `cd backend && nix build .\#backend`
      - With nix without flakes
             - `nix-build -A packages.x86_64-linux.backend-api -o result-backend-api`
             - `nix-build -A packages.x86_64-linux.backend -o result-backend-api`
             - `cd backend-api && ../result-backend-api/bin/backend-api`
             - `cd backend && ../result-backend/bin/backend`

With these steps you are using the compiled js files, but it takes some time to compile. So for quick frontend development I use the following steps:
- Build the vendor-lib files (optional)
       - With flakes: `nix build .\#vendor-lib -o result-vendor`
       - Without flakes: `nix-build packages.x86_64-linux.vendor-lib -o result-vendor`
- Serve the vendor-lib files on port 8081 (optional): `http-server --cors localhost`
- Run backend-api as before
- Run backend as before (optional)
- Enter the develpment shell (`nix-shell` or `nix develop`)
- Enter the repl `cd frontend && cabal repl` 
- Run `mainWithHead` from the repl

In this case we kind of replace `backend` by `frontend` and `http-server` for the static part. 

It might still be tricky to set it up, but I hope you are able to get it running now, and otherwise please let me know.

**Comment by ibizaman** [comment]
I'm writing stuff as I found them.
On the first step, running `nix develop`, I stumbled on
```
error: hash mismatch in fixed-output derivation '/nix/store/vky5s27gyylil96p51is4simphrqnfl6-channel-rust-nightly.toml.drv':
         specified: sha256-B3T/bVER66CE2z7Ocvk7Gn6hXwDZnW+5n/3C0H+HSOk=
            got:    sha256-2023f77rrlQgqax6UkIIjwHip3XSpSkzWuvmhmWWRM0=
```

The fix was easy:
```diff
modified   flake.nix
@@ -86,7 +86,7 @@
         rust_overlay = final: prev: {
           moz-rust = ((prev.rustChannelOf {
             channel = "nightly";
-            sha256 = "B3T/bVER66CE2z7Ocvk7Gn6hXwDZnW+5n/3C0H+HSOk=";
+            sha256 = "2023f77rrlQgqax6UkIIjwHip3XSpSkzWuvmhmWWRM0=";
           }).rust.override { targets = [ "wasm32-unknown-unknown" ]; });
           naersk-lib = naersk.lib."${system}".override {
             cargo = final.moz-rust;
```

I think this will continue to happen if we keep on relying on nightly. Should we continue to do that? It's a good idea to keep up-to-date but I wonder if there's way to do that without making users stumble into this error.

**Comment by stefanboere** [comment]
Thanks! I was expecting that this hash would be no problem, because nixpkgs-mozzilla was pinned in flake.lock, but apparently not. But anyway, rust is only needed for truck-param, and that should really be a separate repo, so I removed this part from this repo now.
